pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java

```diff
@@ -91,6 +95,11 @@ public class DictionaryIndexType
   private static final Logger LOGGER = LoggerFactory.getLogger(DictionaryIndexType.class);
   private static final List<String> EXTENSIONS = Collections.singletonList(V1Constants.Dict.FILE_EXTENSION);
 
+  // Map containing columnName as key and the interner as value. The interner is common across all the segments.
+  // TODO: Check if tablename can be appended to the columnName.
+  private static Map<String, FALFInterner<String>> _strInternerInfoMap = new ConcurrentHashMap<>();
+  private static Map<String, FALFInterner<byte[]>> _byteInternerInfoMap = new ConcurrentHashMap<>();
```

Comment:
I don't think this is the place to store the intern table. I can see several issues:
1. By design, the IndexType should ideally be stateless. It represents the concept of an index. Adding state feels like breaking the separation of concerns.
2. We should try to avoid having mutable state as static variables. I'm not sure how to do that right now, but it is not the first time we find this problem. We really need some server context to store server state.
   * Static mutable state open the door to multiple problems. The most trivial the side effects of running tests. Each test will be affected by the previous ones given these maps will be shared.
4. The structure itself is problematic. Here we are indexing by `columnName`, ignoring the fact that two different tables may have columns with the same name. That is not a correctness issue, but seems strange we want to classify by columnName to improve the cache efficiency but at the same time ignoring the collisions between two different tables.

As said, I don't have a solution right now to be able to implement this without static variables. But what I would suggest is to move this logic to its own singleton class. That at least would solve the separation of concerns issue and we could add a method to that singleton to clean the caches without making the index type more complex than it should.

Reply:
I'm thinking about this and one way to remove the static is to follow the current Pinot design pattern where these kind of structures have a lifetime bound to the query that use them. Specifically, we could keep these two maps in the index reader itself.

This has an obvious disadvantage: there would be one interner per segment and query, so a single query touching two segments or two different queries touching the same segment will not share memory. But it would be a more conservative approach to what we do right now and we could extend the interner to be shared between segments in the future.

Reply:
Good point about keeping DictionaryIndexType stateless. I've moved the static interner maps to a singleton class.

To keep the interners separate for similar column names across different tables, I've created an identifier uses both table name and column name.